### ISSUE TITLE
Wrong control flow order with Constant lookup and ||=

### ISSUE THREAD

**Issue by MaxLap** [oracle/truffleruby#1489]
One of the two issues reported here https://github.com/jruby/jruby/issues/4807 is affecting TruffleRuby.

The issue about `::X ||=` being weird does not affect TruffleRuby.
The other ones does. Here is one of the examples showing off the problem.

The following should puts 'h1' and then fail. Instead, we get "h1" then "h3" then "h1" then fail.
```
(puts 'hi1'; raise 'thing'; puts 'h2';String)::Foo ||= puts 'hi3'
```

This is the commit that solved the issues for JRuby: https://github.com/jruby/jruby/commit/8dae5af511b49742eb5af398197844f4b556579f

**Comment by chrisseaton** [comment]
Wow that's crazy that we have a basic control-flow bug.

**Comment by gogainda** [comment]
Another example: 
```ruby
module Module1
    Var = nil
end

@c = 0
def get_module
	@c += 1
	Module1
end	

(get_module)::Var ||= 42
puts @c
```
on truffleruby it's 3, should be 1

**Comment by chrisseaton** [comment]
Sorry yeah we should get on and this this.

**Comment by norswap** [comment]
Fixed by b981115aa6a3532870c8c0c82f8e90f757185844
